### PR TITLE
Conditionally fence FAMOs under message-order-fence

### PR DESCRIPTION
**Pull request by jhh67**
Previously, to ensure correct AMO ordering in MCM mode message-order-fence, we fenced all fetching atomics (FAMOs) when the provider does not provide `FI_ORDER_ATOMIC_RAW`. This is simple, but has too much overhead. Now we only fence a FAMO if it is necessary to enforce the MCM.

Resolves https://github.com/Cray/chapel-private/issues/5530.

### CODE REVIEW

File: runtime/src/comm/ofi/comm-ofi.c [hunk 1]
```diff
@@ -743,17 +743,20 @@ bitmapBaseType_t bitmapElemBit(size_t i) {
...
 static inline
 void bitmapSet(struct bitmap_t* b, size_t i) {
+  assert(b);
   b->map[bitmapElemIdx(i)] |= bitmapElemBit(i);
 }
 
 static inline
 int bitmapTest(struct bitmap_t* b, size_t i) {
-  return (b->map[bitmapElemIdx(i)] & bitmapElemBit(i)) != 0;
+  return (b && (b->map[bitmapElemIdx(i)] & bitmapElemBit(i)) != 0);
```
Comment (jabraham17) [inline runtime/src/comm/ofi/comm-ofi.c:759]: Its not totally clear to me why these changes where needed? When will `b` ever be `NULL`? Can all of these cases be `assert(b)` (like `bitmapSet`)?

**Review by jabraham17 — APPROVED** [review]
Looks good!

File: runtime/src/comm/ofi/comm-ofi.c (hunk 1 above)
Comment (jhh67) [reply runtime/src/comm/ofi/comm-ofi.c:759]: There is a lot of this idiom in the code:

```
    havePutsOut = (tcip->putVisBitmap != NULL
                   && bitmapTest(tcip->putVisBitmap, node));

```
I put the test for NULL into the function so I could clean iup my code:

```
  chpl_bool havePutsOut = bitmapTest(tcip->putVisBitmap, ab->node);
```

I plan to go back and clean up all the bitmap usage but will do that in a separate PR.

Comment (jhh67) [reply runtime/src/comm/ofi/comm-ofi.c:759]: I added the `assert` to `bitmapSet` because I'm paranoid.
